syft/messaging/plan/plan.py

```diff
@@ -331,6 +354,47 @@ def run(self, args: Tuple, result_ids: List[Union[str, int]]):
             return responses[0]
         return responses
 
+    def has_args_fulfilled(self):
+        """ Check if all the arguments of the plan are ready or not.
+        It might be the case that we still need to wait for some arguments in
+        case some of them are Promises.
+        """
+        for arg_id in self.procedure.arg_ids:
+            arg = self.owner._objects[arg_id]
+            if hasattr(arg, "child") and isinstance(arg.child, PromiseTensor):
+                if not arg.child.is_kept():
+                    return False
+        return True
+
+    def setup_plan_with_promises(self, *args):
+        """ Slightly modifies a plan so that it can work with promises.
+        The plan will also be sent to location with this method.
+        """
+        prom = None
+        for p in args:
+            if hasattr(p, "child") and isinstance(p.child, PromiseTensor):
+                p.child.plans.add(self.id)
+                if prom is None:
+                    prom = p
+
+        # TODO how to infer result type? torch.result_type() in torch 1.3?
+        result_type = args[0].torch_type()  # NOTE torch specfific here
```

Comment:
can't we find another strategy to get the type which is torch agnostic?

Reply:
I forgot this comment, @LaRiffle !
I'm not sure if this is a problem here because PromiseTensors are torch specific